[ch47158] Release/1.0

DamienJacquemart opened the pull request:
- All function in modules under python-lib
- improved tokenizing with 38 dependent-tokenizer modules from Spacy (should be leveraged for other NLP plugins)
- UX harmonized with other NLP plugins
- language detection with cld3 instead of langid

DamienJacquemart committed:
Added all files

- UX is harmonized with other NLP plugins
- all functions are in modules under python-lib
- remaining: 
    - modify language detection lib
    - change tokenizer

DamienJacquemart committed:
Small changes

DamienJacquemart committed:
Update new list

DamienJacquemart committed:
Update recipe.json

DamienJacquemart committed:
Update recipe.py

DamienJacquemart committed:
Update language_detection.py

DamienJacquemart committed:
Update spell_checker.py

DamienJacquemart committed:
Update text_preprocessing.py

alexcombessie committed:
Move plugin files at top level

alexcombessie committed:
Update plugin.json

alexcombessie committed:
renaming recipe folder after adding nlp prefix to plugin id

alexcombessie committed:
Merge branch 'release/1.0' of https://github.com/dataiku/dss-plugin-nlp-spell-checker into release/1.0

alexcombessie committed:
recipe folder renaming

alexcombessie committed:
Python "black" autoformatting

alexcombessie commented on code-env/python/desc.json:
```diff
@@ -0,0 +1,8 @@
+{
+    "acceptedPythonInterpreters": [
+        "PYTHON36"
```
```suggestion
        "PYTHON36", "PYTHON35"
```

alexcombessie commented on code-env/python/spec/requirements.txt:
```diff
@@ -0,0 +1,20 @@
+symspellpy==6.5.2
+spacy>=2.2.2
```
Use == for versions, or >=X.x,<Y.y
```suggestion
spacy==2.3.0
```

alexcombessie commented on code-env/python/spec/requirements.txt:
```diff
@@ -0,0 +1,20 @@
+symspellpy==6.5.2
+spacy>=2.2.2
+nltk==3.4.5
```
nltk doesn't seem to be used, so you can remove it

alexcombessie commented on code-env/python/spec/requirements.txt:
```diff
@@ -0,0 +1,20 @@
+spacy>=2.2.2
+nltk==3.4.5
+jieba==0.42.1
+mecab-python3==0.996.5
```
```suggestion
mecab-python3==1.0.0
```

alexcombessie commented on code-env/python/spec/requirements.txt:
```diff
@@ -0,0 +1,20 @@
+pymorphy2-dicts==2.4.393442.3710985
+pymorphy2-dicts-ru==2.4.404381.4453942
+pymorphy2-dicts-uk==2.4.1.1.1460299261
+pythainlp==2.2.0b1
```
```suggestion
pythainlp==2.2.1
```

alexcombessie replied:
@DamienJacquemart Can you use the latest version? 2.2.2? It's usually not a good idea to specify beta version (...b1) but to always use stable versions.

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+// This file is the descriptor for the Custom code recipe typo-and-lemma
```
Now that the plugin is more complete, you can remove the // comments from the template

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        // description: longer string to help end users understand what this recipe does
+        "description": "Recipe to correct misspelled words in text. <br> Language is either given or detected.",
+        // icon: must be one of the FontAwesome 3.2.1 icons, complete list here at https://fontawesome.com/v3.2.1/icons/
+        "icon": "icon-file-text-alt"
```
```suggestion
        "icon": "icon-strikethrough"
```

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+    // Meta data for display purposes
+    "meta": {
+        // label: name of the recipe as displayed, should be short
+        "label": "Spell Checker",
```
```suggestion
        "label": "Misspelling Correction",
```

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        // label: name of the recipe as displayed, should be short
+        "label": "Spell Checker",
+        // description: longer string to help end users understand what this recipe does
+        "description": "Recipe to correct misspelled words in text. <br> Language is either given or detected.",
```
```suggestion
        "description": "Identify and correct misspellings in text columns for 38 languages",
```

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        {
+            "name": "input_dataset",
+            "label": "Input dataset",
+            "description": "The input dataset from which you want to correct misspelled words.",
```
Small nitpick, we avoid using "." if it's not a valid sentence.
```suggestion
            "description": "Dataset with text columns containing misspelled words",
```

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        },
+        {
+            "name": "not_to_be_corrected_dataset",
+            "label": "Optional: Words not to be corrected",
```
```suggestion
            "label": "Custom vocabulary (optional)",
```

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        {
+            "name": "not_to_be_corrected_dataset",
+            "label": "Optional: Words not to be corrected",
+            "description": "Dataset with a single column of words that will not be corrected.",
```
```suggestion
            "description": "Dataset with a single column of words that should not be corrected",
```

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        {
+            "name": "output_dataset",
+            "label": "Output dataset",
+            "description": "The output dataset with clean text column.",
```
```suggestion
            "description": "Output dataset with identified misspellings and corrected text",
```

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+
+       For the full list and for more details, see the documentation: https://doc.dataiku.com/dss/latest/plugins/reference/params.html
+    */
+    "paramsPythonSetup": "compute_available_time_slots.py",
```
I believe this is a leftover from a copy-paste. Should we remove?

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+            "columnRole": "input_dataset"
+        },
+        {
+            "name": "suffix",
```
I don't think the user should fiddle with the suffix. I would rather we stay in control and hardcode it to "spellchecker". The output columns would be "spellchecker_{}_mispellings", "spellchecker_{}_mispellings_count", "spellchecker_{}_correction".

DamienJacquemart replied:
Word segmentation gives misspelling counts but not the misspelling. I propose to output only 2 columns in that case: "spellchecker{}mispellings_count", "spellchecker{}_correction".

alexcombessie replied:
Okay!

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+            "type": "COLUMNS",
+            "description": "",
+            "mandatory": true,
+            "columnRole": "input_dataset"
```
```suggestion
            "columnRole": "input_dataset",
            "allowedColumnTypes": [
                "string"
            ]
```

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        "type": "SELECT",
+        "name": "language",
+        "label": "Language",
+        "selectChoices": [{'label': 'Albanian', 'value': 'sq'},
```
Could you pass it to a JSON linter?

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+            "defaultValue": "no_typos"
+        },
+        {
+        "type": "SELECT",
```
3 parameters for language handling is a lot. I would favor a single SELECT parameter, with selectChoices in this style:
[
{"label": "Auto-detect (single language)", "value": "auto_single"},
{"label": "Auto-detect (multiple languages)", "value": "auto_multiple"},
{"label": "Albanian", "value": "sq"},
...
]

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+                         {'label': 'Turkish', 'value': 'tr'},
+                         {'label': 'Ukrainian', 'value': 'uk'},
+                         {'label': 'Vietnamese', 'value': 'vi'}],
+        "defaultValue": "en",
```
👍  English is a good default choice

DamienJacquemart replied:
I would propose "detect value", otherwise people will not be aware of the functionality

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        "visibilityCondition": "!model.detect_language"
+        },
+        
+        {
```
to merge in a single SELECT param

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        {
+        "type": "BOOLEAN",
+        "name": "word_segmentation",
+        "label": "Word Segmentation",
```
```suggestion
        "label": "Word segmentation",
```

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        "name": "word_segmentation",
+        "label": "Word Segmentation",
+        "defaultValue": false,
+        "description": "Insert missing spaces at the appropriate positions and correct misspelled words. Slower.",
```
This wording makes me think the plugin will not auto-correct if I don't activate it. Could you clarify?

DamienJacquemart replied:
Je propose:
In addition to correcting misspelled words, word segmentation inserts missing spaces at the appropriate positions (often caused by Typing errors). Slower.

alexcombessie replied:
What about "Detect mangled words and insert missing spaces as required. Slower." ?

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+
+    ],
+
+    // The field "resourceKeys" holds a list of keys that allows to limit the number
```
remove comments

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        "name": "ignore_token",
+        "label": "Ignore token",        
+        "defaultValue": "",
+        "description": "A regex pattern describing what words/phrases to ignore and leave unchanged. Ex: r\"\\w+\\d\" leaves unchanged \"member1\".",
```
```suggestion
        "description": "Regular expression pattern for words/phrases not to be corrected. Ex: r\"^#\\w+" for hashtags.",
```

DamienJacquemart replied:
Bonne idée pour les hashtags !
Par contre, je mettrais r\"#\\w+\" (sans le ^)

alexcombessie replied:
Bien vu

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.json:
```diff
@@ -0,0 +1,196 @@
+        "type": "STRING",
+        "name": "ignore_token",
+        "label": "Ignore token",        
+        "defaultValue": "",
```
```suggestion
        "defaultValue": "r\"^#\\w+",
```

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.py:
```diff
@@ -0,0 +1,22 @@
+import dataiku
+from dataiku.customrecipe import *
```
Use explicit imports instead of *

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.py:
```diff
@@ -0,0 +1,22 @@
+from spell_checker.input_output_params import *
+from spell_checker.spell_checker import *
+
+logger = logging.getLogger(__name__)
```
custom logging won't be used, DSS always overrides it

alexcombessie commented on custom-recipes/nlp-spell-checker-correction/recipe.py:
```diff
@@ -0,0 +1,22 @@
+
+# --- Run
+df = dataiku.Dataset(input_dataset).get_dataframe()
+spell_chckr = SpellChecker(params)
```
nitpick, but I prefer variables names in valid English

alexcombessie commented on python-lib/spell_checker/dataframe_helpers.py:
```diff
@@ -0,0 +1,50 @@
+        return col_name + "_" + suffix
+
+
+def create_all_new_column_names(df_columns, text_col_list):
```
We'll need to add a column with the extracted misspellings (as a list), and a column with the count of mispellings

alexcombessie commented on python-lib/spell_checker/dataframe_helpers.py:
```diff
@@ -0,0 +1,50 @@
+        col_txt_spellchecked_dict[col_txt_name] = new_col
+        df_all_columns.append(new_col)
+
+    return df_all_columns, col_txt_lang_dict, col_txt_preprocessed_dict, col_txt_spellchecked_dict
```
Can you simplify the output to just return a dataframe? Multiple return types makes the code complex to understand and debug.

Possibly, you could define a "formatter class".

alexcombessie commented on python-lib/spell_checker/input_output_params.py:
```diff
@@ -0,0 +1,110 @@
+import dataiku
+from dataiku.customrecipe import *
```
explicit imports instead of star imports

alexcombessie commented on python-lib/spell_checker/input_output_params.py:
```diff
@@ -0,0 +1,110 @@
+
+
+def get_spell_checker_params(recipe_config, not_to_be_corrected_dataset):
+    def _p(param_name, default=None):
```
Use `get` directly, makes the code more readable

alexcombessie commented on python-lib/spell_checker/input_output_params.py:
```diff
@@ -0,0 +1,110 @@
+    params = {}
+
+    # General parameters
+    params["text_col_list"] = _p("text_col_list")
```
Streamlined version: hold your list of parameter names in a CONSTANT, and iterate over the list to assign it to a param dictionary

alexcombessie commented on python-lib/spell_checker/input_output_params.py:
```diff
@@ -0,0 +1,110 @@
+    params["single_language_per_column"] = _p("single_language_per_column")
+    params["constrain_languages"] = _p("constrain_languages")
+
+    if params["detect_language"]:
```
To be simplified (see my comment in recipe.json)

alexcombessie commented on python-lib/spell_checker/input_output_params.py:
```diff
@@ -0,0 +1,110 @@
+
+    # Get set of untouched words
+    if not_to_be_corrected_dataset.is_present:
+        params["set_untouched_words"] = list(not_to_be_corrected_dataset.get_df().iloc[:, 0])
```
This probably needs a preliminary check that the dataset has rows, and a single column. Else, raise an Exception with a custom error mesage.

Oh, just saw that you take care of that in `NotToBeCorrectedDataset`. Good.

alexcombessie commented on python-lib/spell_checker/input_output_params.py:
```diff
@@ -0,0 +1,110 @@
+    """
+
+    if len(ignore_token) > 0:
+        ignore_token = ignore_token[2:-1]  # remove leading r" and ending "
```
In that case, could we tell the user to directly input regex without leading r" and ending "? In all cases, we should be more validation and custom error messaging so the user knows what syntax to use.

alexcombessie commented on python-lib/spell_checker/language_detection.py:
```diff
@@ -0,0 +1,19 @@
+import cld3
+
+
+def get_language_of_text(txt):
```
to be replaced by spacy-fastlang once we have the validation from Clement on the license

alexcombessie commented on python-lib/spell_checker/language_detection.py:
```diff
@@ -0,0 +1,19 @@
+    # the df[col_lang] if filled with the language name
+    if recipe_params["detect_language"]:
+        if recipe_params["single_language_per_column"]:
+            # check language on the 10000 first cells of the text column
```
We shouldn't impose a limit here

alexcombessie commented on python-lib/spell_checker/language_detection.py:
```diff
@@ -0,0 +1,19 @@
+            lang = get_language_of_text(str(df_col_txt[:10000]))
+            return [lang] * nb_rows
+        else:
+            return df_col_txt.apply(lambda x: get_language_of_text(x[:500]))
```
no limit :)

alexcombessie commented on python-lib/spell_checker/spell_checker.py:
```diff
@@ -0,0 +1,127 @@
+import dataiku
+from symspellpy.symspellpy import SymSpell, Verbosity
+from spell_checker.language_detection import *
```
use explicit imports

alexcombessie commented on python-lib/spell_checker/spell_checker.py:
```diff
@@ -0,0 +1,127 @@
+        self.params = params
+
+        # Root path of frequency dictionaries
+        self.resource_path = dataiku.customrecipe.get_recipe_resource()
```
Resource path should be a variable defined outside of the class. As much as possible, we should decorrelate classes from the Dataiku API, so that the code can be ported and tested elsewhere.

alexcombessie commented on python-lib/spell_checker/spell_checker.py:
```diff
@@ -0,0 +1,127 @@
+        self.resource_path = dataiku.customrecipe.get_recipe_resource()
+
+        # sym_spell params. See https://symspellpy.readthedocs.io/en/latest/api/symspellpy.html#symspellpy
+        self.distance = 2
```
This should be an "Expert" parameter which the user can tune

alexcombessie commented on python-lib/spell_checker/spell_checker.py:
```diff
@@ -0,0 +1,127 @@
+
+        # sym_spell params. See https://symspellpy.readthedocs.io/en/latest/api/symspellpy.html#symspellpy
+        self.distance = 2
+        self.suggestion_verbosity = Verbosity.TOP
```
I would suggest an Expert parameter as well

alexcombessie commented on python-lib/spell_checker/spell_checker.py:
```diff
@@ -0,0 +1,127 @@
+
+        for language in self.all_languages:
+            freq_dict_path = self.resource_path + "/" + language + "_50k.txt"
+            sym_spells[language] = SymSpell(max_dictionary_edit_distance=2)
```
need to use class attribute to control edit_distance

alexcombessie commented on python-lib/spell_checker/spell_checker.py:
```diff
@@ -0,0 +1,127 @@
+            return " ".join(sent_fixed)
+        else:
+            try:
+                # for some reason, word_segmentation got a stopIteration error. This is a symspellpy package issue.
```
This seems to be a problem with the dictionary: https://github.com/mammothb/symspellpy/issues/76
Could you investigate to make sure this is not an upstream error with dictionary loading we should fix?

alexcombessie commented on python-lib/spell_checker/spell_checker.py:
```diff
@@ -0,0 +1,127 @@
+                return sym_spell.word_segmentation(
+                    string_or_token_list, ignore_token=self.params["ignore_token"]
+                ).corrected_string
+            except:
```
always catch specific Exception classes. 
```suggestion
            except StopIteration:
                logging.warn("custom error message")
```

alexcombessie commented on python-lib/spell_checker/spell_checker.py:
```diff
@@ -0,0 +1,127 @@
+            self.col_txt_lang_dict,
+            self.col_txt_preprocessed_dict,
+            self.col_txt_spellchecked_dict,
+        ) = create_all_new_column_names(df.columns, self.params["text_col_list"])
```
Hard to understand, this probably needs to be its own method

alexcombessie commented on python-lib/spell_checker/spell_checker.py:
```diff
@@ -0,0 +1,127 @@
+        for col_txt, col_spellchecked in self.col_txt_spellchecked_dict.items():
+            col_lang = self.col_txt_lang_dict[col_txt]
+            col_preprocessed = self.col_txt_preprocessed_dict[col_txt]
+            df[col_spellchecked] = df.apply(
```
small enhancement suggestion: install tqdm and instead progress_apply instead of apply. This will give you a neat progress bar. Great for big datasets!

alexcombessie commented on python-lib/spell_checker/text_preprocessing.py:
```diff
@@ -0,0 +1,101 @@
+import spacy.lang
+
+# All languages available in Spacy and for which we have a frequency dictionnary
+lang_dict = {
```
UPPERCASE for constants

```suggestion
SUPPORTED_LANGUAGES = {
```

alexcombessie commented on python-lib/spell_checker/text_preprocessing.py:
```diff
@@ -0,0 +1,101 @@
+import re
```
remove `re` if unused

alexcombessie commented on python-lib/spell_checker/text_preprocessing.py:
```diff
@@ -0,0 +1,101 @@
+
+class PreprocessText:
+    def __init__(self, all_languages, params=None):
+        self.all_languages = all_languages
```
shouldn't be in the class since it's a constant

alexcombessie commented on python-lib/spell_checker/text_preprocessing.py:
```diff
@@ -0,0 +1,101 @@
+class PreprocessText:
+    def __init__(self, all_languages, params=None):
+        self.all_languages = all_languages
+        if params != None:
```
no need for this, just `self.params=params`

alexcombessie commented on python-lib/spell_checker/text_preprocessing.py:
```diff
@@ -0,0 +1,101 @@
+
+        for lang_code, lang_name in lang_dict.items():
+
+            if lang_code not in self.all_languages:
```
We should logging.warn if the language code is not supported, and make sure the output is empty

alexcombessie commented on python-lib/spell_checker/text_preprocessing.py:
```diff
@@ -0,0 +1,101 @@
+        return tokenizers
+
+    def _tokenize(self, doc, tokenizer):
+        return [str(k) for k in tokenizer(doc)]
```
Use https://spacy.io/api/tokenizer#pipe (will be much faster)

alexcombessie commented on python-lib/spell_checker/text_preprocessing.py:
```diff
@@ -0,0 +1,101 @@
+        print("Pre-processing text ...")
+        for col_txt, col_preprocessed in col_txt_preprocessed_dict.items():
+            col_lang = col_txt_lang_dict[col_txt]
+            df[col_preprocessed] = df.apply(
```
progress_apply from tqdm (cf previous comment)

alexcombessie commented on python-lib/spell_checker/spell_checker.py:
```diff
@@ -0,0 +1,127 @@
+from spell_checker.text_preprocessing import *
+
+
+class SpellChecker:
```
A suggestion: The "words not to be corrected" (to be renamed to Custom Vocabulary) could be added to the dictionaries. The small complexity would be recomputing term frequencies for these words.

I let you evaluate if this should be in v1 or in v2. It's a nice to have.

alexcombessie reviewed (commented):
Thanks for the good work @DamienJacquemart! I have left a sequence of comments inline.

DamienJacquemart committed:
Update code-env/python/desc.json

Co-authored-by: Alex Combessie <alex.combessie@gmail.com>

DamienJacquemart committed:
Update code-env/python/spec/requirements.txt

small fixes on versions

Co-authored-by: Alex Combessie <alex.combessie@gmail.com>

DamienJacquemart committed:
remove nltk

DamienJacquemart committed:
small fixes

DamienJacquemart committed:
remove // from template

DamienJacquemart commented on plugin.json:
```diff
@@ -1,29 +1,15 @@
-    "version": "0.0.1",
-
-    // Meta data for display purposes:
+    "id": "nlp-spell-checker",
```
How to change the id of a plugin ? (initally "spell-checker")

alexcombessie commented:
Clubhouse integration: [ch47158](https://app.clubhouse.io/dataiku/story/47158/autocorrection-plugin-recipe)

DamienJacquemart committed:
First version without language detection

alexcombessie committed:
Update plugin.json

alexcombessie committed:
reinit branch to empty state

DamienJacquemart committed:
Improved pre-processing
Removed language detection

alexcombessie commented on code-env/python/spec/requirements.txt:
```diff
@@ -0,0 +1,16 @@
+SudachiDict-core==20200330
+natto-py==0.9.2
+pymorphy2-dicts-uk==2.4.1.1.1460299261
+git+https://github.com/kmike/pymorphy2.git@ca1c13f6998ae2d835bdd5033c17197dcba84cf4
```
@DamienJacquemart Can you point to pypi? https://pypi.org/project/pymorphy2/
It's usually not a good idea to use git install in plugin requirements if there is a pypi alternative.

DamienJacquemart replied:
That does not work for Ukrainien. You can also go with the git. It's the workaround given by pymorph2.

alexcombessie replied:
I am not sure to fully understand the issue, and couldn't find it in the doc or github. Do you have a link to investigate?

DamienJacquemart replied:
This is what you get when pip installing

pymorphy2==0.8
pymorphy2-dicts==2.4.393442.3710985
pymorphy2-dicts-ru==2.4.404381.4453942
pymorphy2-dicts-uk==2.4.1.1.1460299261

Job failed: Error in Python process: At line 28: <class 'ImportError'>: The Ukrainian lemmatizer requires the pymorphy2 library and dictionaries: try to fix it with "pip uninstall pymorphy2" and"pip install git+https://github.com/kmike/pymorphy2.git pymorphy2-dicts-uk"

DamienJacquemart replied:
I removed Ukrainian support.

alexcombessie commented on custom-recipes/nlp-spell-checker/recipe.json:
```diff
@@ -0,0 +1,157 @@
+            "type": "SEPARATOR"
+        },
+        {
+        "type": "SELECT",
```
@DamienJacquemart can you use the same pattern as here: https://github.com/dataiku/dss-plugin-amazon-comprehend-nlp/blob/master/custom-recipes/amazon-comprehend-nlp-named-entity-recognition/recipe.json?

alexcombessie commented on custom-recipes/nlp-spell-checker/recipe.json:
```diff
@@ -0,0 +1,157 @@
+        "type": "SELECT",
+        "name": "language_from_list",
+        "label": "Language",
+         "selectChoices":[{'label': 'Albanian', 'value': 'sq'},
```
@DamienJacquemart Can you lint the JSON? (" instead of ', apply automatic formatting)

DamienJacquemart replied:
Just for the information, will there be an issue somewhere using " instead of ' ?

alexcombessie commented on custom-recipes/nlp-spell-checker/recipe.json:
```diff
@@ -0,0 +1,157 @@
+        "name": "ignore_token",
+        "label": "Ignore token",        
+        "defaultValue": "",
+        "description": "Regular expression pattern for words/phrases not to be corrected. Ex: \\w+\\d will preserve \"home1\" ",
```
@DamienJacquemart I remembered we said we would use the example of the regex for @ mentions

DamienJacquemart replied:
hashtags and mentions are removed by the tokenizer. So they are lost using symspell.

alexcombessie replied:
I understand that, but the \w+ example isn't the most appropriate in my opinion. In the context of spelling correction, you would actually not use \w+ as an ignore pattern. That would mean no words are corrected. I think we should find a better example, which would be valid in the context of spelling correction.

DamienJacquemart replied:
What about "containing both letters and number", as an ID would do ?

alexcombessie replied:
Yes, that would be clearer. I would also add somewhere that hashtags and mentions are directly removed, so that the user knows he doesn't need to take care of that.

DamienJacquemart replied:
We should handle this more carefully. 
-> Hashtags and mentions can be great in a NLP pipeline: wordcloud, classification, ...
-> keeping them and not treating would make the code slower
-> not doing anything would lead to correct the hashtags into words of the dictionary
Any thoughts on this ?

alexcombessie replied:
You raise a good point. For spell checker is OK to remove hashtags and mentions. But for other plugins, it's going to be a problem. 
Is it possible to prevent the tokenizer from doing this filtering by default? 
And then in that case, we can use the ignore_token parameter as intended, to remove them?

DamienJacquemart replied:
Yes, we should do this. I put it on my to do.
In case it adds a lot of computation time, we could set this as an option.

alexcombessie replied:
Good idea

alexcombessie commented on custom-recipes/nlp-spell-checker/recipe.py:
```diff
@@ -0,0 +1,36 @@
+params = load_plugin_config(get_recipe_config(), custom_vocabulary_set)
+
+# --- Run
+df = input_dataset.get_dataframe()
```
We shouldn't load the full dataset in memory, since we are using chunk processing. I assume this is a leftover from old code, can we remove?
```suggestion
```

DamienJacquemart replied:
absolutely !

alexcombessie commented on custom-recipes/nlp-spell-checker/recipe.py:
```diff
@@ -0,0 +1,36 @@
+                folder_of_dictionaries = params['folder_of_dictionaries']
+                )
+
+output_df = spell_checker_object.compute(df)
```
Using process_dataset_chunks instead
```suggestion
```

alexcombessie commented on custom-recipes/nlp-spell-checker/recipe.py:
```diff
@@ -0,0 +1,36 @@
+
+# --- Run
+df = input_dataset.get_dataframe()
+spell_checker_object = SpellChecker(
```
@DamienJacquemart is this fast enough to be made into a processor? Or should we keep it recipe only?

DamienJacquemart replied:
It takes a long time to load tokenizer and symspell objects. For one language it is acceptable, otherwise not.

alexcombessie replied:
Okay, let's drop it then.

alexcombessie commented on python-lib/plugin_config_loading.py:
```diff
@@ -0,0 +1,94 @@
+
+from language_dict import SUPPORTED_LANGUAGES
+
+def load_plugin_config(recipe_config: Dict, custom_vocabulary_set: List) -> Dict:
```
@DamienJacquemart good job! This is great!

DamienJacquemart replied:
thank you, you taught them :-)

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+    # See https://symspellpy.readthedocs.io/en/latest/api/symspellpy.html#symspellpy
+    DISTANCE = 2 # distance max under which a correction is returned
+    SUGGESTION_VERBOSITY = Verbosity.TOP # returns only the closest word
+    TRANSFERT_CASING = False # the original casing (lowercase and uppercase) is not carried over the text. Symspellpy returns errors for some words. Note that if a word is all uppercase, symspellpy returns the lowercased word. 
```
small typo
```suggestion
    TRANSFER_CASING = False # the original casing (lowercase and uppercase) is not carried over the text. Symspellpy returns errors for some words. Note that if a word is all uppercase, symspellpy returns the lowercased word. 
```

DamienJacquemart replied:
I did not have the plugin finished at this stage, the code might still have typos ^^

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+    """
+    
+    # See https://symspellpy.readthedocs.io/en/latest/api/symspellpy.html#symspellpy
+    DISTANCE = 2 # distance max under which a correction is returned
```
This should be a parameter in the recipe

DamienJacquemart replied:
agree

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+    
+    def __init__(self, 
+                text_column_list: List = [],
+                language_selection: AnyStr = "from_column",
```
Just keep the language and language_column parameter

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+        if self.language_selection == 'from_list':
+            self.language = language
+            
+        self.supported_lang_code = SUPPORTED_LANGUAGES.keys()
```
Use the constant directly (otherwise this hides the fact that it's a constant)

alexcombessie commented on custom-recipes/nlp-spell-checker/recipe.json:
```diff
@@ -0,0 +1,157 @@
+        {
+            "name": "text_column_list",
+            "label": "Text columns",
+            "type": "COLUMNS",
```
For consistency with other plugins, I think we should keep it to one column. This will simplify the code a lot.

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+    TRANSFERT_CASING = False # the original casing (lowercase and uppercase) is not carried over the text. Symspellpy returns errors for some words. Note that if a word is all uppercase, symspellpy returns the lowercased word. 
+    
+    def __init__(self, 
+                text_column_list: List = [],
```
I would keep it to one column, for simplicity and consistency with other plugins.

DamienJacquemart replied:
Changed to have only one column.

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+        for lang_code in new_lang_code_list:
+       
+            if lang_code not in self.supported_lang_code:
+                continue
```
logging.warning

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+        # sym spell objects as dictionary lang_id: sym_spell_object
+        self.sym_spells = {}   
+            
+    def _add_sym_spell_objects(self, new_lang_code_list: List):
```
why new_lang_code_list?

DamienJacquemart replied:
a new list of potential codes is given at each chunk of data

alexcombessie replied:
Okay, gotcha.

How about we have a dedicated method to initialize and load language dictionary in the beginning, using a `dataset.get_dataframe(columns=[language_column])`?

DamienJacquemart replied:
Could be, I though it would be slower. I guess for a .csv, it has to read the entire file to get the column, write ?

alexcombessie replied:
You have a good point. My proposed solution would preserve memory, but would be slow. Thinking about it, your solution, loading at each chunk, is more efficient. My only concern would be: does it imply to load the same dictionary several times? Ideally, we could load dictionaries for the first chunk, then for other chunks, we only load dictionaries that are not already present.

DamienJacquemart replied:
The class SymSpell and TextPreprocess are instantiated before the chunk processing. dict and tokenizer of chunks are added only if they are not already present from previous chunks.

alexcombessie replied:
Okay in that case we're good. Would be great to have function docstrings and more comments, as it was not very simple to understand what was going on.

DamienJacquemart replied:
done

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+                continue
+            logging.info("Loading sym spell object for language {}".format(lang_code))
+            freq_dict_path = self.folder_of_dictionaries + '/' + lang_code + ".txt"
+            self.sym_spells[lang_code] = SymSpell(max_dictionary_edit_distance=2)
```
max_dictionary_edit_distance should be a parameter in the UI

DamienJacquemart replied:
done

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+            
+        else:
+            corrected_word = word # if no correction was found, original word is returned
+            misspell = word # if no correction was found, we assume the word is misspelled
```
If the word is correct, it doesn't count, right?

DamienJacquemart replied:
If no correction was found, the distance between the misspelled word and the words from the dictionary is higher than "max_edit_distance". In that case, we assume the word is misspelled. 

The case where the correction is the same as the (lowercase version) of the original word is handle just before in

if correction[0].term != word.lower() and word != ' ':

alexcombessie replied:
gotcha

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+        """
+
+        if lang not in self.supported_lang_code:
+            return ('', '', '')
```
logging.warning

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+        
+        return (corrected_sentence, misspellings, misspelling_count)
+            
+    def compute(self, df: pd.DataFrame) -> pd.DataFrame:
```
This shouldn't process the entire dataframe, since we're using chunk processing.
We probably need to do something to initialize the list of languages at the beginning.

In general, this method is too long, you need to break it down into smaller ones

DamienJacquemart replied:
Happy to discuss this.

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+        
+        # Fix typos
+        logging.info("Fixing typos ...")
+        for txt_col in self.text_column_list:        
```
Let's start with one column only first.

DamienJacquemart replied:
done

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+        logging.info("Fixing typos ...")
+        for txt_col in self.text_column_list:        
+            
+            # Name creation of new columns
```
This should be simpler as a for loop

DamienJacquemart replied:
done: only one column

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+                                                       lang_col)
+
+            # Add new sym_spell objects
+            new_lang_code_list = list(df[lang_col].unique())
```
This should be a separate method

DamienJacquemart replied:
happy to discuss this.

alexcombessie commented on python-lib/text_preprocessing.py:
```diff
@@ -0,0 +1,131 @@
+        
+        for lang_code in lang_code_new_list:
+            
+            if lang_code != lang_code: # check for NaNs
```
Not the most understandable solution to check for NaNs, I would rather test that len(lang_code) is 2

DamienJacquemart replied:
I thought None could be given by some DB, and None has no length. Do you infirm this ?

alexcombessie replied:
I am not sure. How about we "validate" language_code with `lang_code in SUPPORTED_LANGUAGE_CODE`, else log a warning. That would cover all invalid cases.

Actually, in the case of tokenization, the case where lang_code not in SUPPORTED_LANGUAGE_CODE should just pass to a spacy nlp object without specific language

DamienJacquemart replied:
Not sure to understand: you want to raise only one warning each time an encontered lang code is unsupported ?

what would be the benefit to add in a spacy nlp object without specific language ?

alexcombessie replied:
Yes, one warning per unsupported language code.
spacy nlp has a default tokenizer which is not language specific. This tokenization library should be as generic and reusable as possible, so have this default tokenizer as a fallback is a good thing.

alexcombessie commented on python-lib/text_preprocessing.py:
```diff
@@ -0,0 +1,131 @@
+        for lang_code in lang_code_new_list:
+            
+            if lang_code != lang_code: # check for NaNs
+                logging.info("Missing language code")
```
```suggestion
                logging.warning("Missing language code")
```

alexcombessie commented on python-lib/text_preprocessing.py:
```diff
@@ -0,0 +1,131 @@
+                # new tokenizer is added only if not already present
+                continue
+            
+            # Special treatment for Korean as Spacy korean tokenizer has dependecies
```
Can you detail what is the problem with Korean? Ideally, we should use spacy directly

DamienJacquemart replied:
You must install system dependencies on the server in order to use Spacy.

alexcombessie replied:
Do you have a link? I am not sure to understand the issue.

DamienJacquemart replied:
At the bottom of this [page](https://spacy.io/usage/models#languages), you see Korean dependencies. [natto-py](https://github.com/buruzaemon/natto-py) needs mecab, that needs extra install on the server.

alexcombessie replied:
Good catch. It seems this problem is also highlighted among spacy developers (https://github.com/explosion/spaCy/issues/5708)

So yes, until spaCy figures out Korean support without system dependency, let's use your solution :)

alexcombessie commented on python-lib/text_preprocessing.py:
```diff
@@ -0,0 +1,131 @@
+                nlps[lang_code] = getattr(language_modules[lang_code], lang_name)()
+                self.tokenizers[lang_code] = nlps[lang_code].Defaults.create_tokenizer(nlps[lang_code])
+                
+    def _first_preprocessing(self, doc: AnyStr,
```
I am not sure to understand why this method is needed?

DamienJacquemart replied:
lowercase: if a word is all lowercase, symspell returns a mix of capital and lowercase letter in a word
remove punctuation: symspell corrects punctuation into a word

alexcombessie replied:
Gotcha. I was confused by the method name. I would suggest a more understandable name like _normalize_text, and adding a function documentation string to explain why this method is needed.

DamienJacquemart replied:
done

alexcombessie commented on code-env/python/spec/requirements.txt:
```diff
@@ -0,0 +1,16 @@
+git+https://github.com/kmike/pymorphy2.git@ca1c13f6998ae2d835bdd5033c17197dcba84cf4
+pythainlp==2.2.0b1
+konlpy==0.5.2
+
```
We also need to bundle the spacy language-specific models: https://spacy.io/usage/models#models-download

DamienJacquemart replied:
We can pip install them also.

alexcombessie replied:
yes, pip is the way to go

DamienJacquemart replied:
Where this is needed in the code actually ?

alexcombessie replied:
all spacy models for tokenization

DamienJacquemart replied:
hum. I do not remembered have seen pip install capabilities for all the models. Do you have this ?

alexcombessie replied:
For English, it's https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-2.2.0/en_core_web_sm-2.2.0.tar.gz#egg=en_core_web_sm
(I am quoting https://spacy.io/usage/models#models-download)
For other languages, they can all be found in https://github.com/explosion/spacy-models/releases/

alexcombessie commented on python-lib/text_preprocessing.py:
```diff
@@ -0,0 +1,131 @@
+                continue
+                
+            if lang_code not in self.supported_lang_code:
+                logging.info("Unsupported language code {}".format(lang_code))
```
logging.warning

alexcombessie commented on python-lib/text_preprocessing.py:
```diff
@@ -0,0 +1,131 @@
+                lang_name = SUPPORTED_LANGUAGES[lang_code]
+
+                # module import
+                logging.info("Loading tokenizer object for language {}".format(lang_code))
```
This should use language models bundled in code-env

alexcombessie commented on python-lib/text_preprocessing.py:
```diff
@@ -0,0 +1,131 @@
+                                               axis=1)
+
+        # tokenize
+        df[preprocess_col] = df.apply(lambda x:self._tokenize(x[first_preprocessing_col],
```
This should use https://spacy.io/usage/processing-pipelines it will be infinitely faster

DamienJacquemart replied:
happy to discuss this

alexcombessie commented on python-lib/text_preprocessing.py:
```diff
@@ -0,0 +1,131 @@
+        existing_column_names = list(df.columns)
+        first_preprocessing_col = generate_unique(txt_col, existing_column_names, 'edge_case')
+        
+        df[first_preprocessing_col] = df.apply(lambda x:self._first_preprocessing(x[txt_col],
```
Not sure why this should be assigned to df since we don't keep it

DamienJacquemart replied:
for the next apply. Is there a better way to do ?

alexcombessie replied:
Okay, I would keep it simple and do one big apply.

alexcombessie commented on python-lib/plugin_config_loading.py:
```diff
@@ -0,0 +1,94 @@
+    # Language selection
+    params["language_selection"] = recipe_config.get("language_selection")
+    
+    if params["language_selection"] == 'from_list':
```
I would simplify that a little bit to keep only a `language` and a `language_column` parameter

DamienJacquemart replied:
done

alexcombessie commented on python-lib/plugin_config_loading.py:
```diff
@@ -0,0 +1,94 @@
+        else:
+            params["ignore_token"] = recipe_config.get("ignore_token")
+            # Check for valid regex
+            try:
```
I would keep it simple, no need for try/except. If it fails, the error message is clear enough, correct?

DamienJacquemart replied:
right.

DamienJacquemart replied:
euh actually I do not see a better way to do it

alexcombessie replied:
Just use the command without the try except. If re.compile fails, it will send an exception by itself.

DamienJacquemart replied:
I would keep the assert.
The error from re.compile is less explicit. Tried with the regex '[', this give:
Job failed: Error in Python process: At line 14: <class 'sre_constants.error'>: unterminated character set at position 0 More info about this error

alexcombessie replied:
Ah yes that's definitely not a good error message. Let's not change then.

alexcombessie commented on python-lib/plugin_config_loading.py:
```diff
@@ -0,0 +1,94 @@
+            logging.info("Token pattern to be ignored {}".format(params["ignore_token"]))
+
+    else:
+        logging.info("Expert mode is disabled")
```
I wouldn't change the behavior depending on the activation of expert mode. For me, expert mode should just be applied for visibility.
For instance, if a user activates expert mode, adds an ignore_token, and then deactivates expert mode, we should keep the ignore_token.

DamienJacquemart replied:
I believe this is very confusing, as you can mask the expert mode and it is still activated. But you are the plugin guy :-)

alexcombessie replied:
Yes, but to change it in the first place, the users needs to activate expert mode :) If the user changed something, we shouldn't change it "behind his back". 

Another way to put it is the frontend shouldn't make backend decisions. Expert mode is purely a frontend thing. Whether the user displays expert parameters or not shouldn't change the way the frontend sends data to the backend. It's purely a visibility thing.

DamienJacquemart replied:
ok. Done.

alexcombessie commented on python-lib/text_preprocessing.py:
```diff
@@ -0,0 +1,131 @@
+
+from language_dict import SUPPORTED_LANGUAGES
+
+class TextPreprocessor:
```
As a general comment, this class could use a bit of reorganization: small number of methods, clearer argument names, comments, function documentation strings.

I have made a few comments inline.

DamienJacquemart replied:
happy to further discuss.

alexcombessie commented:
@DamienJacquemart I just finished the 2nd review and left comments inline. 

We are mostly good on the UX, although we can simplify language choosing and add a parameter for the user to control the correction max edit distance.

On the backend, there are several important simplifications and optimizations to bring. In particular, the spaCy tokenization needs proper bundling and pipelining. We also need to dig to the bottom of the issue with Korean. It's important we get the tokenization library right, as it will be used by all other plugins.

For the autocorrection part, it seems mostly OK from a code perspective (only a few enhancement suggestions). However, I have a question on dictionaries: Are we using https://github.com/wolfgarbe/SymSpell#frequency-dictionaries-in-other-languages or only https://github.com/hermitdave/FrequencyWords?

alexcombessie reviewed (changes requested):
2nd review ongoing, cf last messages

alexcombessie commented on python-lib/spell_checker.py:
```diff
@@ -0,0 +1,197 @@
+        correction = self.sym_spells[lang].lookup(word, 
+                                                  self.SUGGESTION_VERBOSITY, 
+                                                  max_edit_distance=self.DISTANCE, 
+                                                  transfer_casing=self.TRANSFERT_CASING,
```
```suggestion
                                                  transfer_casing=self.TRANSFER_CASING,
```

DamienJacquemart committed:
Update code-env/python/spec/requirements.txt

Co-authored-by: Alex Combessie <alex.combessie@gmail.com>

DamienJacquemart committed:
fix pythainlp version

DamienJacquemart committed:
small changes

DamienJacquemart committed:
MultiThread for spell_checker.py

Added column description

nlp object added as explained here (begining of the nlp pipeline)
https://spacy.io/usage/processing-pipelines#processing

Removed ukrainian support

DamienJacquemart committed:
Tokenizer handles email, url, hashtags, mentions (only hashtag need a custom tokenizer)
New functions created in text_preprocessing to check for email, url, hashtags and mentions

Tokenization into a nlp object (begining of a SpaCy pipeline)

Removed Ukrainian
Removed Korean (too many special cases to handle, let's reserve it for a V2)
-> Now hanldes 36 languages instead of 38

Column description in the output dataset

DamienJacquemart committed:
Documentation improved of text_preprocessing.py

Removing '.' and '/' with a seprate function in text_preprocessing.py

DamienJacquemart committed:
Small enhancements

DamienJacquemart committed:
mini modif

alexcombessie committed:
Code polishing

DamienJacquemart committed:
- Added emoji except zh, th, ja, see https://github.com/PyThaiNLP/pythainlp/issues/465

- disable tagger, parser

- fixed bug with url

- punctuation is retrieved, but it adds extra spaces. ('hellu.' -> 'hello .') No idea how do to better.

- fixed "hello." now it is tokenized ["hello", "."]

- hello.goodbye is still one token hello.goodbye which is spacy behaviour

- dictionnary has not "'nt", "'d", ... hence there are considered as spekking mistakes. We can add them to the dictionary.

DamienJacquemart committed:
small enhancements

DamienJacquemart committed:
small enhancement

alexcombessie committed:
UX and code enhancements

alexcombessie committed:
first pass at generic multilingual tokenizer object

alexcombessie committed:
Polishing the precious tokenizer

alexcombessie committed:
work in progress on spell_checker refactoring

alexcombessie committed:
module renaming

alexcombessie committed:
typo fix on tokenizer

alexcombessie committed:
better error message

alexcombessie committed:
good progress on spellchecker

alexcombessie committed:
fully working version

alexcombessie committed:
batch_size for tokenizer

alexcombessie committed:
Fastaaa!!

alexcombessie committed:
better, faster, somewhat stronger

alexcombessie committed:
a few enhancements

alexcombessie committed:
edge punctuation case

alexcombessie committed:
better edge case handling

alexcombessie committed:
add feature: custom corrections input

alexcombessie committed:
finishing touches

alexcombessie committed:
Code documentation ongoing

alexcombessie committed:
More docstring

alexcombessie committed:
Fix additional edge cases

alexcombessie committed:
better logging for words non corrected

alexcombessie committed:
Add option to use spaCy models

alexcombessie committed:
error handling of model loading

alexcombessie committed:
Can't use statistical NER to improve correction

alexcombessie committed:
Ongoing work on diagnosis output

alexcombessie committed:
Test commit

alexcombessie committed:
diagnosis output work in progress

alexcombessie committed:
tokenizer can haz lemma

alexcombessie committed:
small improvement of tokenizer, making sure it can be operate with scikit-learn

alexcombessie committed:
docstring, docstring

alexcombessie committed:
I can haz docstring

alexcombessie committed:
edge case with default tokenizer not entirely compatible with spacymoji

alexcombessie committed:
deactivate multithreading when counting tokens

alexcombessie committed:
small enhancements

alexcombessie committed:
Misc improvements:
- bump version of symspellpy
- newer english dictionary
- error handling of sympsell (edge case)
- custom exception instead of assertion errors for param validation
- improve logging messages

alexcombessie committed:
remove spacymoji

alexcombessie committed:
start working on unit tests

alexcombessie reviewed (commented):
Resolved comments with latest commits

alexcombessie commented:
@StanislasGuinel I let you review the PR again, as I brought quite a few changes following my previous review

alexcombessie committed:
working unit-test for tokenizer

alexcombessie committed:
fix unit tests

alexcombessie committed:
work on unit test for spellchecker

alexcombessie committed:
working spellchecker test

alexcombessie committed:
Update recipe.json

alexcombessie committed:
Update plugin_io_utils.py

alexcombessie committed:
Update plugin_io_utils.py

StanislasGuinel commented on python-lib/plugin_config_loading.py:
```diff
@@ -0,0 +1,174 @@
+    # List of text columns
+    params["text_column"] = recipe_config.get("text_column")
+    logging.info("Text column: {}".format(params["text_column"]))
+    if params["text_column"] == "":
```
Need to check that it is a valid column name and also check for both "" and None with "not"
```suggestion
    if not params["text_column"]:
```

StanislasGuinel commented on python-lib/plugin_config_loading.py:
```diff
@@ -0,0 +1,174 @@
+            raise PluginParamValidationError("Empty language column selection")
+        logging.info("Language column: {}".format(params["language_column"]))
+    else:
+        if params["language"] is None or params["language"] == "":
```
```suggestion
        if not params["language"]:
```

StanislasGuinel commented on python-lib/plugin_config_loading.py:
```diff
@@ -0,0 +1,174 @@
+    if col_type != "string":
+        raise PluginParamValidationError("Column of custom vocabulary dataset must be of string type")
+
+    df = clean_text_df(custom_vocabulary_dataset.get_dataframe())
```
why not infer_with_pandas=False like in the custom_corrections_checker function ?

alexcombessie replied:
Good catch, yes
```suggestion
    df = clean_text_df(custom_vocabulary_dataset.get_dataframe(infer_with_pandas=False))
```

alexcombessie replied:
Solved with my last commit 524ffe5

StanislasGuinel commented on python-lib/plugin_io_utils.py:
```diff
@@ -0,0 +1,119 @@
+    Returns:
+       pandas.DataFrame with rows dropped in case of empty strings or NaN values
+    """
+    for col in df.columns:
```
why doing it on al columns and not only dropna_columns ?
is it because then you do .fillna("").astype(str) to the output in custom_corrections_checker() ?

alexcombessie replied:
I need to do `replace("", np.NaN)` first to use `dropna`

StanislasGuinel commented on python-lib/plugin_io_utils.py:
```diff
@@ -0,0 +1,119 @@
+    if dropna_columns is None:
+        df = df.dropna()
+    else:
+        df = df.dropna(subset=dropna_columns)
```
subset can be None
```suggestion
    df = df.dropna(subset=dropna_columns)
```

StanislasGuinel commented on python-lib/dku_io_utils.py:
```diff
@@ -0,0 +1,108 @@
+                if output_dataset.writePartition is None or output_dataset.writePartition == "":
+                    output_dataset.write_schema_from_dataframe(output_df, dropAndCreate=True)
+                else:
+                    output_dataset.write_schema_from_dataframe(output_df)
```
just a suggestion :)
```suggestion
               output_dataset.write_schema_from_dataframe(output_df, dropAndCreate=bool(not output_dataset.writePartition)
```

StanislasGuinel commented on python-lib/dku_io_utils.py:
```diff
@@ -0,0 +1,108 @@
+        Number of records
+    """
+    metric_id = "records:COUNT_RECORDS"
+    dataset_name = dataset.name.split(".")[1]
```
```suggestion
    dataset_name = dataset.short_name
```

StanislasGuinel commented on python-lib/symspell_checker.py:
```diff
@@ -0,0 +1,437 @@
+        Args:
+            df: Input pandas DataFrame
+        """
+        del df[self._tokenizer.tokenized_column]
```
it seems to be the recommended way of removing columns in pandas
```suggestion
       df.drop(self._tokenizer.tokenized_column, axis=1, inplace=True)
```

StanislasGuinel commented on custom-recipes/nlp-spell-checker/recipe.py:
```diff
@@ -0,0 +1,39 @@
+set_column_description(
+    input_dataset=params["input_dataset"],
+    output_dataset=params["output_dataset"],
+    column_description_dict=spellchecker._output_column_description_dict,
```
```_output_column_description_dict``` should not be a private field

alexcombessie replied:
You're right, let me correct that

alexcombessie replied:
Solved with my last commit 524ffe5

StanislasGuinel commented on python-lib/dku_io_utils.py:
```diff
@@ -0,0 +1,108 @@
+
+
+def set_column_description(
+    output_dataset: dataiku.Dataset, column_description_dict: Dict, input_dataset: dataiku.Dataset = None
```
```column_description_dict``` you should not put the type in the variable name

alexcombessie replied:
In general, I think this is a good rule. However, there are cases where I find that writing the type in the variable name is helpful to other developers and avoid mistakes. The `:` type hint is not always easy to spot when writing code which works on top of existing code. If I were to apply this rule everywhere, then I should rename `output_dataset` to `output` and `input_dataset` to `input`, which would make the code harder to understand.

In this particular case, I am not sure it's straightforward to understand that the chosen data structure for `column_description` is a `Dict`. I am using variants of column_description_dict throughout my code, and this function is used by several plugins. So I am not sure that we would gain from renaming this particular variable.

StanislasGuinel replied:
good point !

StanislasGuinel commented on python-lib/spacy_tokenizer.py:
```diff
@@ -0,0 +1,300 @@
+        """
+        start = time()
+        logging.info("Loading tokenizer for language '{}'...".format(language))
+        if language in SPACY_LANGUAGE_MODELS.keys() and self.use_models:
```
```suggestion
        if language in SPACY_LANGUAGE_MODELS and self.use_models:
```

StanislasGuinel commented on python-lib/spacy_tokenizer.py:
```diff
@@ -0,0 +1,300 @@
+        added_tokenizer = False
+        if pd.isnull(language) or language == "":
+            raise ValueError("Missing language code")
+        if language not in SUPPORTED_LANGUAGES_SPACY.keys():
```
```suggestion
        if language not in SUPPORTED_LANGUAGES_SPACY:
```

StanislasGuinel commented on python-lib/spacy_tokenizer.py:
```diff
@@ -0,0 +1,300 @@
+            raise ValueError("Missing language code")
+        if language not in SUPPORTED_LANGUAGES_SPACY.keys():
+            raise ValueError("Unsupported language code: '{}'".format(language))
+        if language not in self.spacy_nlp_dict.keys():
```
```suggestion
        if language not in self.spacy_nlp_dict:
```

StanislasGuinel commented on python-lib/spacy_tokenizer.py:
```diff
@@ -0,0 +1,300 @@
+        self.use_models = use_models
+        self.hashtags_as_token = hashtags_as_token
+        self.batch_size = int(batch_size)
+        self.spacy_nlp_dict = {}
```
should not write type in variable name

alexcombessie replied:
I have the same comment as for `column_description_dict`. In this particular case, it's not straightforward that this particular object is a `Dict`. This class is meant to be used by a lot of other classes and other plugins. Even though it is not very elegant, I feel that for this particular object, including dict in the name can avoid confusions from other developers integrating this class.

StanislasGuinel replied:
alright !

StanislasGuinel reviewed (commented):
First review (not finished), I did many tests also.
The code is very clean and well documented, not many comments.

alexcombessie committed:
Apply suggestions from code review

Co-authored-by: StanislasGuinel <60091564+StanislasGuinel@users.noreply.github.com>

alexcombessie committed:
enhancements after code review

alexcombessie committed:
Update python-lib/symspell_checker.py

Co-authored-by: StanislasGuinel <60091564+StanislasGuinel@users.noreply.github.com>

alexcombessie committed:
Fix identation error

StanislasGuinel commented on python-lib/symspell_checker.py:
```diff
@@ -0,0 +1,438 @@
+                3. Spellchecker diagnosis string explaining the spellchecker action
+        """
+        (is_misspelled, correction, diagnosis) = (False, token.text, "")
+        if token.text in self.custom_corrections.keys():  # special case of custom corrections
```
```suggestion
        if token.text in self.custom_corrections:  # special case of custom corrections
```

StanislasGuinel commented on python-lib/symspell_checker.py:
```diff
@@ -0,0 +1,438 @@
+                document_slice = df.loc[language_indices, self._tokenizer.tokenized_column]  # slicing df by language
+                if len(document_slice) != 0:
+                    tuple_list = self.check_document_list(document_list=document_slice, language=lang)
+                    for i, column in enumerate(self.output_column_description_dict.keys()):
```
```suggestion
                    for i, column in enumerate(self.output_column_description_dict):
```

StanislasGuinel commented on python-lib/symspell_checker.py:
```diff
@@ -0,0 +1,438 @@
+                        )
+        else:
+            tuple_list = self.check_document_list(document_list=df[self._tokenizer.tokenized_column], language=language)
+            for i, column in enumerate(self.output_column_description_dict.keys()):
```
```suggestion
            for i, column in enumerate(self.output_column_description_dict):
```

StanislasGuinel commented on python-lib/symspell_checker.py:
```diff
@@ -0,0 +1,438 @@
+        """
+        df = pd.DataFrame()
+        logging.info("Computing spellchecker diagnosis...")
+        for i, column in enumerate(self.DIAGNOSIS_COLUMN_DESCRIPTION_DICT.keys()):
```
```suggestion
        for i, column in enumerate(self.DIAGNOSIS_COLUMN_DESCRIPTION_DICT):
```

StanislasGuinel commented on python-lib/plugin_config_loading.py:
```diff
@@ -0,0 +1,174 @@
+    params["language"] = recipe_config.get("language")
+    if params["language"] == "language_column":
+        params["language_column"] = recipe_config.get("language_column")
+        if params["language_column"] is None or params["language_column"] == "":
```
```suggestion
        if not params["language_column"]:
```

StanislasGuinel commented on python-lib/plugin_config_loading.py:
```diff
@@ -0,0 +1,174 @@
+    if params["language"] == "language_column":
+        params["language_column"] = recipe_config.get("language_column")
+        if params["language_column"] is None or params["language_column"] == "":
+            raise PluginParamValidationError("Empty language column selection")
```
check it is a valid column from the input dataset

alexcombessie replied:
Good point, solved by 3c0cf4c

StanislasGuinel commented on python-lib/plugin_config_loading.py:
```diff
@@ -0,0 +1,174 @@
+    # List of text columns
+    params["text_column"] = recipe_config.get("text_column")
+    logging.info("Text column: {}".format(params["text_column"]))
+    if not params["text_column"]:
```
check also it is a valid column from the input dataset

alexcombessie replied:
Good point, solved by 3c0cf4c

StanislasGuinel reviewed (commented):
some more small fixes

StanislasGuinel committed:
check python3 version in makefile

alexcombessie committed:
Apply suggestions from code review

remove unnecessary `.keys()` when iterating over a dictionary

Co-authored-by: StanislasGuinel <60091564+StanislasGuinel@users.noreply.github.com>

alexcombessie committed:
add check that text column and language column do exist

alexcombessie committed:
- Adjustements on the makefile
- Fix spellchecker unit tests

alexcombessie committed:
Merge pull request #3 from dataiku/fix/makefile-updated

check python3 version in makefile

alexcombessie committed:
makefile typo

alexcombessie commented:
@StanislasGuinel I passed a few small changes, I let you tell me if it's good to go when you come back from vacation!
